File: app/src/main/java/com/example/personalaifoodmap/MainActivity.java
package com.example.personalaifoodmap;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class MainActivity extends AppCompatActivity {

    int foodGalleryMode = 1;

    public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button allImageBtn = (Button)findViewById(R.id.allImageBtn);
        allImageBtn.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                foodGalleryMode = 1;
                Intent intent = new Intent(view.getContext(), GalleryActivity.class);
                intent.putExtra("isFoodGallery", foodGalleryMode);
                view.getContext().startActivity(intent);
            }
        });

        Button foodImageBtn = (Button)findViewById(R.id.foodImageBtn);
        foodImageBtn.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                foodGalleryMode = 2;
                Intent intent = new Intent(view.getContext(), GalleryActivity.class);
                intent.putExtra("foodGalleryMode", foodGalleryMode);
                view.getContext().startActivity(intent);
            }
        });

        Button foodMapBtn = (Button)findViewById(R.id.foodMapBtn);
        foodMapBtn.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View view) {
                foodGalleryMode = 3;
                Intent intent = new Intent(getApplicationContext(),  GalleryActivity.class);
                intent.putExtra("foodGalleryMode", foodGalleryMode);
                view.getContext().startActivity(intent);
            }
        });

    }
}

File: app/src/main/java/com/example/personalaifoodmap/MainActivity.kt
package com.example.personalaifoodmap

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.drawable.Drawable
import android.media.ExifInterface
import android.net.Uri
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.text.TextUtils
import android.view.View
import android.widget.Button
import android.widget.ImageView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import kotlinx.android.synthetic.main.activity_main.*
import org.jetbrains.anko.alert
import org.jetbrains.anko.noButton
import org.jetbrains.anko.toast
import org.jetbrains.anko.yesButton
import java.io.File
import java.io.FileOutputStream
import javax.sql.DataSource

class MainActivity : AppCompatActivity() {

    private val REQUEST_READ_EXTERMAL_STORAGE = 1000

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // 권한 부여 확인
        if(ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)!= PackageManager.PERMISSION_GRANTED){
            // 권한 거부 시
            if(ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.READ_EXTERNAL_STORAGE)){
                // 이전에 이미 권한이 거부되었을 때
                alert("외부 저장소 권한이 필요합니다") {
                    yesButton {
                        // 권한 요청
                        ActivityCompat.requestPermissions(this@MainActivity,
                            arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE),
                            REQUEST_READ_EXTERMAL_STORAGE)
                    }
                    noButton {  }
                }.show()
            } else {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE),
                    REQUEST_READ_EXTERMAL_STORAGE)
            }
        } else {
            // 권한이 이미 허용됨
            getAllPhotos()

        }
    }

    // 사용자가 권한 요청 시 호출되는 메서드
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when(requestCode){
            REQUEST_READ_EXTERMAL_STORAGE -> {
                if((grantResults.isNotEmpty() && grantResults[0]
                            == PackageManager.PERMISSION_GRANTED)) {
                    getAllPhotos()
                } else {
                    toast("권한 거부")
                }
                return
            }
        }
    }

    private fun getAllPhotos(){
        val cursor = contentResolver.query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
            null,
            null,
            null,
            MediaStore.Images.ImageColumns.DATE_TAKEN + " DESC")
        val uriArr = ArrayList<String>()
        if(cursor!=null){
            while(cursor.moveToNext()){
                // 사진 경로 Uri 가져오기
                val uri = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA))
                uriArr.add(uri)
            }
            cursor.close()
        }
        val adapter = MyAdapter(this,uriArr)
        gridView.numColumns=3 // 한 줄에 3개씩 사진 출력
        gridView.adapter = adapter

    }
}
